### PR TITLE
Fix return type in arange and zero step size handling.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

File: numba/targets/arrayobj.py [hunk 1]
```diff
@@ -3765,7 +3765,23 @@ def _arange_dtype(*args):
...
+        #
+        # It's therefore not possible to replicate the case where integer
+        # literals are supplied, however best effort is made to correctly handle
+        # cases like:
+        # * np.arange(np.int8(10)) -> max(np.long, np.int8) -> np.long
+        #
+        # Alg ref:
+        # https://github.com/numpy/numpy/blob/maintenance/1.17.x/numpy/core/src/multiarray/ctors.c#L3376-L3377    # noqa: E501
+        dtype = max(bounds + [types.long_,])
+        print(bounds, "resolved", dtype, "long = ", types.long_)
```
Comment (seibert) [inline numba/targets/arrayobj.py]: Leftover print statement to remove here.

Comment (stuartarchibald) [reply numba/targets/arrayobj.py]: tests are failing, I put that in to see what was going on :)

**Comment by seibert** [comment]
linux 32-bit seems to still be unhappy.

File: numba/targets/arrayobj.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/targets/arrayobj.py]: I thought it was this: https://github.com/numba/numba/pull/4882/files#diff-81d8f85ead1fb3ae948c94ed8d2a969eR3768-R3782 but it seems like a `types.long_` is appearing as an `int64` on linux32 and NumPy wants an `int32`.
https://dev.azure.com/numba/numba/_build/results?buildId=3317&view=logs&jobId=536d9229-f68f-5476-bdce-7191d78ec342&taskId=695d2d88-17fa-5407-392b-a8e99ffa20e2&lineStart=317&lineEnd=321&colStart=1&colEnd=1

Comment (seibert) [reply numba/targets/arrayobj.py]: ah, makes sense

**Comment by seibert** [comment]
Azure has passed, and this looks like a reasonable way to deal with the platform quirks.

**Pull request merged**